### PR TITLE
Upgrade dev dependencies

### PR DESCRIPTION
**Pull request by li-kai**
1. Upgrades babel & jest
2. Improve jest performance
3. Improve babel and js performance

### CODE REVIEW

File: www/babel.config.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  ];
+
+  const plugins = [
+    '@babel/plugin-syntax-dynamic-import',
+    // // See https://github.com/facebook/create-react-app/issues/4263
+    ['@babel/plugin-proposal-class-properties', { loose: true }],
+    ['@babel/plugin-proposal-optional-chaining', { loose: true }],
+  ];
+
+  if (IS_DEV || IS_PROD) {
```
Comment (ZhangYiJiang) [inline www/babel.config.js]: I think we can keep this for tests. Do we need any of the extra lodash functions / features in tests?

File: www/babel.config.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        exclude: ['transform-typeof-symbol'],
+      },
+    ],
+    ['@babel/preset-react', { development: !IS_PROD }],
+    '@babel/preset-flow',
+  ];
+
+  const plugins = [
+    '@babel/plugin-syntax-dynamic-import',
+    // // See https://github.com/facebook/create-react-app/issues/4263
```
Comment (ZhangYiJiang) [inline www/babel.config.js]: /nit double comment (what does it mean!?)

**Comment by ZhangYiJiang** [comment]
Code coverage somehow increased 16%. Either you broke the code coverage path config or you fixed it. Can you check which one it is? :P

**Comment by ZhangYiJiang** [comment]
CI seem to be slightly faster, which is nice, and bundle size is smaller by 4kb gzipped, which is a small win

File: www/package.json [hunk 3]
```diff
@@ -64,6 +63,7 @@
     "eslint-plugin-jsx-a11y": "6.0.3",
     "eslint-plugin-prettier": "2.6.0",
     "eslint-plugin-react": "7.7.0",
+    "eslint": "4.19.1",
```
Comment (taneliang) [inline www/package.json]: Why did `eslint` get moved down?

File: www/babel.config.js (hunk 1 above)
Comment (li-kai) [reply www/babel.config.js]: No, it's just that compilation would be slower if we included it

File: www/package.json (hunk 3 above)
Comment (li-kai) [reply www/package.json]: Sorted by alphabetical order

File: www/babel.config.js (hunk 2 above)
Comment (li-kai) [reply www/babel.config.js]: Added a comment, is that what you were looking for?

Comment (taneliang) [reply www/babel.config.js]: Don't think so. Line 27 has two `//`s haha (`// // `). The new comment is useful though, let's leave it there

**Comment by li-kai** [comment]
@ZhangYiJiang It was broken because it wasn't covering all the files without a test file beside. Now it is still up, because we ignore test-utils instead of test_utils now :100:

File: www/babel.config.js (hunk 1 above)
Comment (ZhangYiJiang) [reply www/babel.config.js]: Probably worth keeping the lodash plugin just to keep test behavior consistent with the other envs

File: www/jest.config.js [hunk 4]
```diff
@@ -1,41 +1,25 @@
...
-    'src/js/test_utils',
-    'src/js/e2e',
-    // Code in this file triggers this bug - https://github.com/istanbuljs/babel-plugin-istanbul/issues/116
-    'src/js/views/modules/ModulePageContent.jsx',
-  ],
+  roots: ['<rootDir>/src/js'],
+  moduleDirectories: ['node_modules', '<rootDir>/src/js'],
+  moduleFileExtensions: ['jsx', 'js'],
+  testPathIgnorePatterns: [],
+  testRegex: 'src/js/.+\\.test\\.jsx?$',
```
Comment (ZhangYiJiang) [inline www/jest.config.js]: Can this be a regex literal instead?

File: www/jest.config.js [hunk 5]
```diff
@@ -1,41 +1,25 @@
...
+    '\\.(?:css|scss)$': 'identity-obj-proxy',
   },
   // Allow us to directly use enzyme wrappers for snapshotting
   // Usage: expect(enzyme.shallow(<div/>)).toMatchSnapshot();
-  snapshotSerializers: [
-    '<rootDir>/node_modules/enzyme-to-json/serializer',
-  ],
+  snapshotSerializers: ['<rootDir>/node_modules/enzyme-to-json/serializer'],
+  collectCoverageFrom: ['**/*.{js,jsx}'],
+  coveragePathIgnorePatterns: ['/node_modules/', '<rootDir>/src/js/(?:test-utils|e2e)'],
```
Comment (ZhangYiJiang) [inline www/jest.config.js]: I think the `?:` is unnecessary. It makes it a non-capturing group, which captures intent better, but it makes it a bit harder to read. I can go either way 🤷‍♂️

File: www/jest.config.js (hunk 4 above)
Comment (li-kai) [reply www/jest.config.js]: Unfortunately, no.
```
Option "testRegex" must be of type:
    string or array
```

**Pull request merged**